src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetAuthenticationCredentialsResponse.cs

```diff
@@ -0,0 +1,83 @@
+using System.Globalization;
+using System.Linq;
+using Newtonsoft.Json;
+
+namespace NuGet.Protocol.Plugins
+{
+    /// <summary>
+    /// A get authentication credentials response
+    /// </summary>
+    public sealed class GetAuthenticationCredentialsResponse
+    {
+        /// <summary>
+        /// Username
+        /// </summary>
+        public string Username { get; }
+
+        /// <summary>
+        /// password token
+        /// </summary>
+        public string Password { get; }
+
+        /// <summary>
+        /// message - optional, can be used as a way to communicate to NuGet why the authentication failed.
+        /// </summary>
+        public string Message { get; }
+
+        /// <summary>
+        /// Gets or sets the list of authentication types this credential is applicable to. Useful values include
+        /// <c>basic</c>, <c>digest</c>, <c>negotiate</c>, and <c>ntlm</c>
+        /// </summary>
+        public IList<string> AuthenticationTypes { get; }
+
+        /// <summary>
+        /// ResponseCode - status of the credentials
+        /// </summary>
+        [JsonRequired]
+        public MessageResponseCode ResponseCode { get; }
+
+        /// <summary>
+        /// Create a response object
+        /// </summary>
+        /// <param name="username"></param>
+        /// <param name="password"></param>
+        /// <param name="message"></param>
+        /// <param name="authenticationTypes"></param>
+        /// <param name="responseCode"></param>
+        /// <exception cref="ArgumentException">If MessageResponseCode is not defined on this runtime</exception>
+        [JsonConstructor]
+        public GetAuthenticationCredentialsResponse(string username, string password, string message, IList<string> authenticationTypes, MessageResponseCode responseCode)
+        {
+
+            if (!Enum.IsDefined(typeof(MessageResponseCode), responseCode))
+            {
+                throw new ArgumentException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        Strings.Plugin_UnrecognizedEnumValue,
+                       nameof(responseCode)));
+            }
+
+
+            Username = username;
+            Password = password;
+            Message = message;
+            AuthenticationTypes = authenticationTypes;
+            ResponseCode = responseCode;
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the provider returnd a valid response.
+        /// </summary>
+        /// <remarks>
+        /// Either Username or Password (or both) must be set, and AuthTypes must either be null or contain at least
+        /// one element
+        /// </remarks>
+        public bool IsValid => ResponseCode == MessageResponseCode.Success && (!string.IsNullOrWhiteSpace(Username) || !string.IsNullOrWhiteSpace(Password))
+                               && (AuthenticationTypes == null || AuthenticationTypes.Any());
```

Comment:
`(AuthenticationTypes == null || AuthenticationTypes.Any())` didn't understand this condition clearly... we dont accept when someone explicitly set this list to empty but fine with null?

Reply:
Yeah, that's exactly what it's supposed to be. 
This is equivalent to our current authentication code. 
If it's not set we create NetworkCredential, if it is, we create AuthTypeFilteredCredentials.

Reply:
It just feels odd, ideally we should just take empty as null as well in this case.